src/pip/_internal/metadata/importlib/_dists.py

```diff
@@ -109,6 +109,17 @@ def from_directory(cls, directory: str) -> BaseDistribution:
         dist = importlib.metadata.Distribution.at(info_location)
         return cls(dist, info_location, info_location.parent)
 
+    @classmethod
+    def from_metadata_file(
+        cls,
+        metadata_path: str,
+        filename: str,
+        project_name: str,
+    ) -> BaseDistribution:
+        metadata_location = pathlib.Path(metadata_path)
+        dist = importlib.metadata.Distribution.at(metadata_location.parent)
```

Comment:
This sounds pretty reasonable, and investigating it led me to realize that the new abstract classmethod `from_metadata_file()` that I added is probably misnamed: the goal of that method is to return a *metadata for a wheel distribution*, but *without* having to provide a zip file. See in particular the pkg_resources impl here and contrast it to `from_directory()` right above it (which constructs a `PathMetadata` instead of a `WheelMetadata`): https://github.com/pypa/pip/blob/a685a984d2faae9a3ffbd4a68db45e9cee84123e/src/pip/_internal/metadata/pkg_resources.py#L95-L112

I was worried that returning a `WheelMetadata` would mean our code would fail for non-wheel dists, but our tests for this feature are actually only for sdists right now (which I will fix!): https://github.com/pypa/pip/blob/a685a984d2faae9a3ffbd4a68db45e9cee84123e/tests/functional/test_download.py#L1271-L1273

So I'm going to make a few changes now:
1. Rename `pip._internal.metadata.pkg_resources.WheelMetadata` to `DictMetadata` since it's being used to provide metadata for more than just wheels now.
1. Rename `pip._internal.metadata.base.BaseDistribution.from_metadata_file()` to `from_metadata_file_contents()`, since it's not supposed to be used to parse any actual metadata directory on disk (this may involve generating a temp dir).
1. Add test cases for `data-dist-info-metadata` which use wheels!

Reply:
Interestingly, `WheelMetadata` _was_ at a point named `DictMetadata`, but renamed since it’s unclear what that class is for (is it a dict that representing the `METADATA` file?) Perhaps `InMemoryMetadata` could be a better alternative.

Reply:
Thanks for doing that spelunking! I have made those renames/logic changes and will now circle back to adding test cases for wheels with `data-dist-info-metadata` entries.